src/System.ServiceModel.Primitives/tests/Channels/CommunicationObjectTest.cs

```diff
@@ -0,0 +1,544 @@
+            mco.DefaultOnOpening();
+        };
+
+        // We don't expect OnOpen to be called using the asynchronous open,
+        // so this override will expose whether that happens
+        mco.OnOpenOverride = (TimeSpan t) =>
+        {
+            openMethodsCalled.Add("OnOpen");
+            mco.DefaultOnOpen(t);
+        };
+
+        mco.OnBeginOpenOverride = (TimeSpan t, AsyncCallback c, object s) =>
+        {
+            openMethodsCalled.Add("OnBeginOpen");
+            return mco.DefaultOnBeginOpen(t, c, s);
+        };
+
+        mco.OnOpenedOverride = () =>
+        {
+            openMethodsCalled.Add("OnOpened");
+            mco.DefaultOnOpened();
+        };
+
+        mco.OnClosingOverride = () =>
+        {
+            closeMethodsCalled.Add("OnClosing");
+            mco.DefaultOnClosing();
+        };
+
+        mco.OnCloseOverride = (TimeSpan t) =>
+        {
+            closeMethodsCalled.Add("OnClose");
+            mco.DefaultOnClose(t);
+        };
+
+        mco.OnBeginCloseOverride = (TimeSpan t, AsyncCallback c, object s) =>
+        {
+            closeMethodsCalled.Add("OnBeginClose");
+            return mco.DefaultOnBeginClose(t, c, s);
+        };
+
+        mco.OnClosedOverride = () =>
+        {
+            closeMethodsCalled.Add("OnClosed");
+            mco.DefaultOnClosed();
+        };
+
+        // *** EXECUTE *** \\
+        IAsyncResult openAr = mco.BeginOpen(TimeSpan.FromSeconds(30), callback: null, state: null);
+        mco.OpenAsyncResult.Complete();
+        mco.EndOpen(openAr);
+
+        IAsyncResult closeAr = mco.BeginClose(TimeSpan.FromSeconds(30), callback: null, state: null);
+        mco.CloseAsyncResult.Complete();
+        mco.EndClose(closeAr);
+
+        // *** VALIDATE *** \\
+        string expectedOpens = "OnOpening,OnBeginOpen,OnOpened";
+        string actualOpens = String.Join(",", openMethodsCalled);
+        Assert.True(String.Equals(expectedOpens, actualOpens, StringComparison.Ordinal),
+               String.Format("Expected open methods to be '{0}' but actual was '{1}'.",
+                             expectedOpens, actualOpens));
+
+        string expectedCloses = "OnClosing,OnBeginClose,OnClosed";
+        string actualCloses = String.Join(",", closeMethodsCalled);
+        Assert.True(String.Equals(expectedCloses, actualCloses, StringComparison.Ordinal),
+               String.Format("Expected close methods to be '{0}' but actual was '{1}'.",
+                             expectedCloses, actualCloses));
+    }
+
+#if FULLXUNIT_NOTSUPPORTED
+    [Fact]
+#endif
+    [WcfFact]
+    public static void CommunicationObject_Abort_Close_Methods_Called()
+    {
+        MockCommunicationObject mco = new MockCommunicationObject();
+        List<string> closeMethodsCalled = new List<string>();
+
+        // *** SETUP *** \\
+        mco.OnClosingOverride = () =>
+        {
+            closeMethodsCalled.Add("OnClosing");
+            mco.DefaultOnClosing();
```

Comment:
I suppose we can have a common setup function that takes a `MockCommunicationObject` argument. It will hook up all the On*Overrides and probably events as well. Each test will call the same setup function so we don't worry about missing anything (also reduced duplicated code), but each test will have different execution block so is validation block.